packages/host/src/brain.rs

```diff
@@ -171,15 +185,21 @@ impl Brain {
 
                 display: Display::new(peripherals_tx.clone(), display_rx),
             }),
+            link_addr: link_addr_2,
         }
     }
 
     pub async fn run_program(
         &mut self,
         mut qemu_command: Command,
         kernel: PathBuf,
-        binary: PathBuf,
+        main_binary: PathBuf,
+        main_binary_addr: u32,
+        linked_binary: Option<PathBuf>,
+        linked_binary_addr: Option<u32>,
     ) -> io::Result<()> {
+        assert!(linked_binary_addr.unwrap_or(1) != 0, "link address cannot be zero!");
```

Comment:
If the `linked_binary_addr` cannot be zero, then this should be an `Option<NonZeroU32>` to enforce that requirement via the type system rather than relying on an undocumented panic.